src/hotspot/share/memory/resourceArea.hpp

```diff
@@ -120,13 +122,28 @@ class ResourceArea: public Arena {
     } else {
       assert(size_in_bytes() == state._size_in_bytes, "Sanity check");
     }
-    _chunk = state._chunk;      // Roll back to saved chunk.
-    _hwm = state._hwm;
-    _max = state._max;
 
-    // Clear out this chunk (to detect allocation bugs)
-    if (ZapResourceArea) {
-      memset(state._hwm, badResourceValue, state._max - state._hwm);
+    if (have_more_chunks || (_hwm != state._hwm)) {
```

Comment:
If we started another chunk, _hwm cannot be state._hwm. So I would shorten this to `if  (_hwm != state._hwm)`. You can add an assert that the hwm should be different if we have follow up chunks.

Reply:
All right, done so.